### PR TITLE
feat: `std/data_structures`

### PR DESCRIPTION
**Pull request by iuioiua**
This change moves the unstable `BinaryHeap()`, `BinarySearchTree()`, `ascend()`, `descend()` and `RedBlackTree()` APIs to the new `std/data_structures` sub-module. After a discussion with @kt3k, we decided this location for these APIs makes the most sense. This will also open up opportunities for more data-structure-related APIs in the future.

The unstable files within `std/collections` were deleted, as they were... unstable. I've updated and cleaned up the previous deprecation documentation and added the sub-module to CI processes, etc.

### CODE REVIEW

**Comment by kt3k** [comment]
Any suggestion, objection, bikeshedding, etc from the community is welcome.

**Comment by iuioiua** [comment]
There seems to be no concern.

File: .github/CODEOWNERS
Comment (bartlomieju) [inline .github/CODEOWNERS:1]: I think we should remove this file altogether (in a separate PR)

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by iuioiua** [comment]
I've corrected the removal version.

**Comment by kt3k** [comment]
Let me try some bike shedding. How about singular form: `std/data_structure`?

@KyleJune Do you have any other/better ideas for this new submodule name?

**Comment by iuioiua** [comment]
I think `data_structures` makes more sense as the folder contains implementations for multiple data structures.

**Comment by KyleJune** [comment]
Of the two, my preference would be std/data_structures to make it clear there are many. std/collections is also plural.

**Review by kt3k — APPROVED** [review]
LGTM 👍

**Pull request merged**